L2ARC: Rework write throttling with DWPD rate limiting and parallel writes

ixhamza opened the pull request:
### Motivation and Context
The L2ARC write throttling algorithm was designed years ago and has several limitations in modern environments:
- **Random sublist selection**: The introduction of ARC multilists made L2ARC writing random, picking one sublist at a time and depending on whether it found something to write there, leading to inconsistent write patterns and uneven cache population.
- **Redundant tail scanning**: Each L2ARC iteration scans from the tail of ARC eviction lists. Since lists are only filled from heads, we repeatedly scan the same already-processed entries, wasting CPU cycles. The scanning depth is limited by a conservative headroom setting to reduce this overhead, but it also limits write throughput.
- **Single feed thread**: A single global feed thread served all L2ARC devices across the entire system sequentially. With multiple pools or multiple cache devices, this serialized all writes and prevented scaling throughput with additional devices.
- **Bursty and unstable rates**: Write patterns are bursty and often don't reach target speed. After burst workloads end, L2ARC writing stops for extended periods, failing to maintain consistent write rates.
- **Obsolete boost mechanism**: `l2arc_write_boost` was designed before persistent L2ARC existed. It tied write rate acceleration to ARC warmup state, but ARC being cold doesn't mean L2ARC should write faster if it's full, and ARC being warm doesn't mean L2ARC should slow down if it's empty.

### Description
- **Even-depth multi-sublist scanning**: Replace random single-sublist selection with systematic scanning across all sublists within each burst. Headroom is distributed evenly across sublists for consistent cache population.
- **Persistent markers**: Save scan positions between iterations using per-sublist markers. This eliminates redundant tail scanning and significantly reduces CPU overhead. Large devices `(capacity >= arc_c_max/2)` use persistent markers; small devices retain original HEAD/TAIL behavior based on ARC warmup state.
- **ABD deferred free fix**: Reorder destruction in `arc_hdr_destroy()` to destroy L1HDR before L2HDR, allowing proper ABD deferral when L2_WRITING flag is set. Prevents gang ABD panics with multiple L2ARC devices due to I/O aggregation reusing ABDs still in gang chains.
- **Per-device feed threads**: Replace single global feed thread with per-device threads. Each L2ARC device runs independently with its own thread, enabling parallel writes. Sublist busy tracking coordinates between threads to prevent simultaneous access to the same sublist. Throughput scales linearly with cache devices.
- **DWPD-based rate limiting**: Replace bursty algorithm with Drive Writes Per Day limiting to protect SSD endurance. New l`2arc_dwpd_limit` tunable (default 1) controls writes per day per device capacity. Removes `l2arc_write_boost` and `l2arc_write_interval()`. Features include:
  - Budget calculated from `device size × DWPD limit`.
  - Unused budget accumulates over 24-hour periods with automatic reset and carry-over.
  - Controlled bursts (max 50MB [`L2ARC_BURST_SIZE_MAX`]) with adaptive intervals to achieve target rates.
  - Applies after initial fill when total L2ARC capacity meets persist threshold (`arc_c_max/2`).
  - First pass writes are free, DWPD budget starts fresh after device fills.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
- New ZTS tests: `l2arc_dwpd_pos`, `l2arc_dwpd_multidev_pos`, `l2arc_dwpd_persist_pos`, `l2arc_parallel_writes_pos`.
- CI Testing
- Manual testing with multiple L2ARC devices

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Quality assurance (non-breaking change which makes the code more robust against bugs)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

DeHackEd commented:
Just as an idea, there are SSDs out there with a DWPD figure of less than 1.0, and people share SSDs for multiple purposes. Maybe the module parameter should be a percentage or other fractional unit?

ixhamza commented:
Ready for review. Updated DWPD parameter to percentage-based for fractional support, [resolved `arc_release()` accounting issue](https://github.com/openzfs/zfs/pull/18093/commits/278ba4d824e76883e1315ca77f626d87ba46d5b4), and rebased to master.

behlendorf commented on include/sys/arc_impl.h:
```diff
@@ -41,14 +41,25 @@
 
+/*
+ * We can feed L2ARC from two states of ARC buffers, mru and mfu,
+ * and each of the state has two types: data and metadata.
```
state -> states

behlendorf commented on tests/zfs-tests/include/tunables.cfg:
Should be removed.

behlendorf commented on tests/zfs-tests/tests/functional/l2arc/l2arc_dwpd_ratelimit_pos.ksh:
```diff
@@ -0,0 +1,124 @@
+typeset noprefetch=$(get_tunable L2ARC_NOPREFETCH)
+typeset dwpd_limit=$(get_tunable L2ARC_DWPD_LIMIT)
+typeset arc_min=$(get_tunable ARC_MIN)
+typeset arc_max=$(get_tunable ARC_MAX)
```
nit: while we don't use them consistently (yet) the `save_tunable` / `restore_tunable` helper functions were added for this.

ixhamza commented on module/zfs/arc.c:
```diff
@@ -9044,48 +9007,139 @@ l2arc_read_done(zio_t *zio)
+			multilist_sublist_unlock(mls);
+		}
+	}
+}
```
False Positive: `KM_SLEEP` allocations cannot fail.

ixhamza commented on module/zfs/arc.c:
```diff
@@ -10032,8 +10399,37 @@ l2arc_add_vdev(spa_t *spa, vdev_t *vd)
+		    spa_name(spa), (u_longlong_t)vd->vdev_id);
+		adddev->l2ad_feed_thread = thread_create_named(thread_name,
+		    NULL, 0, l2arc_feed_thread, adddev, 0, &p0, TS_RUN,
+		    minclsyspri);
```
Not very critical but added.

amotin commented on tests/zfs-tests/tests/functional/l2arc/l2arc_multidev_scaling_pos.ksh:
```diff
@@ -0,0 +1,131 @@
+typeset cache_sz=900   # 900MB per device
+typeset fill_mb=2500   # 2.5GB initial data
+typeset test_time=25   # Measurement window: 16MB/s × 25s = 400MB per device
+typeset VDEV_CACHE2="$VDIR/cache2"
```
I was wonder if this line would look better in `l2arc.cfg`, but I see in the following test you are even more free with the names.

amotin commented on tests/zfs-tests/tests/functional/l2arc/l2arc_multidev_throughput_pos.ksh:
```diff
@@ -0,0 +1,110 @@
+# Verify writes are at least 80% of expected
+typeset min_bytes=$((expected * 80 / 100))
+if [[ $bytes -lt $min_bytes ]]; then
+	log_fail "Writes ${bytes_mb}MB below minimum $((min_bytes/1024/1024))MB"
```
I am somewhat worrying about reliability of benchmarks in several times virtualized test environment, but once we are here, I think we could also have maximum threshold, not only minimum.

amotin commented on module/zfs/arc.c:
```diff
@@ -9179,6 +9168,70 @@ l2arc_log_blk_overhead(uint64_t write_sz, l2arc_dev_t *dev)
+	uint64_t available_budget = dwpd_budget + dev->l2ad_dwpd_accumulated;
+	if (expected_writes > dev->l2ad_dwpd_writes) {
+		/* Add unused budget from current period */
+		available_budget += expected_writes - dev->l2ad_dwpd_writes;
```
You've lost me here.  You first add raw bytes accumulated yesterday to a current bytes per second rate, then conditionally add raw bytes unused today.  Despite different measurement units, the last addition kind of makes sense, allowing almost full speed writes if we have something saved something today.  Same time it seems to me that in case we saved some writes yesterday, then we'll effectively have no DWPD limit today at all, since we'll now have `l2ad_dwpd_accumulated` allowance every second.

ixhamza commented:
@behlendorf - Rebased to master. The only CI failure (FreeBSD 16 `zpool_iostat_interval_all`) is unrelated to this patchset.
